File: fractal_sequences/transformations.py
"""Transform sequence values and attractors into discrete domains."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .coordinates import CoordinateMap, CoordinateSystem
from .discretization import DiscretizationGrid
from .models import Attractor, Number, SequenceAnalysisResult, SequenceTerm


@dataclass(frozen=True)
class DiscretePoint:
    """Result of projecting and discretizing a single value."""

    value: Number
    coordinates: CoordinateMap
    indices: dict[str, int] | None
    discrete_coordinates: CoordinateMap | None
    cell_id: int | None
    inside_domain: bool


@dataclass(frozen=True)
class DiscreteTerm:
    """A sequence term plus its discrete projection."""

    term: SequenceTerm
    point: DiscretePoint


@dataclass(frozen=True)
class DiscreteAttractor:
    """An attractor plus the discrete projection of its values."""

    attractor: Attractor
    points: list[DiscretePoint]


@dataclass(frozen=True)
class DiscreteAnalysisResult:
    """Sequence analysis and its discrete representation."""

    sequence_result: SequenceAnalysisResult
    transformer: "DiscreteTransformer"
    terms: list[DiscreteTerm]
    attractors: list[DiscreteAttractor]
    config: Any | None = None

    @classmethod
    def from_analysis(
        cls,
        sequence_result: SequenceAnalysisResult,
        transformer: "DiscreteTransformer",
        config: Any | None = None,
    ) -> "DiscreteAnalysisResult":
        return cls(
            sequence_result=sequence_result,
            transformer=transformer,
            terms=transformer.transform_terms(sequence_result.terms),
            attractors=transformer.transform_attractors(sequence_result.attractors),
            config=config,
        )


class DiscreteTransformer:
    """Project values into coordinates and quantize them on a uniform grid."""

    def __init__(
        self,
        coordinate_system: CoordinateSystem,
        grid: DiscretizationGrid,
    ) -> None:
        missing_axes = set(grid.axes) - set(coordinate_system.axes)
        if missing_axes:
            missing = ", ".join(sorted(missing_axes))
            raise ValueError(f"coordinate system does not provide axes: {missing}")

        self.coordinate_system = coordinate_system
        self.grid = grid

    def transform_value(self, value: Number) -> DiscretePoint:
        coordinates = self.coordinate_system.coordinates_from(value)
        indices = self.grid.indices_of(coordinates)

        if indices is None:
            return DiscretePoint(
                value=value,
                coordinates=coordinates,
                indices=None,
                discrete_coordinates=None,
                cell_id=None,
                inside_domain=False,
            )

        return DiscretePoint(
            value=value,
            coordinates=coordinates,
            indices=indices,
            discrete_coordinates=self.grid.centers_of(indices),
            cell_id=self.grid.cell_id(indices),
            inside_domain=True,
        )

    def transform_terms(self, terms: list[SequenceTerm]) -> list[DiscreteTerm]:
        return [
            DiscreteTerm(term=term, point=self.transform_value(term.value))
            for term in terms
        ]

    def transform_attractor(self, attractor: Attractor) -> DiscreteAttractor:
        return DiscreteAttractor(
            attractor=attractor,
            points=[self.transform_value(value) for value in attractor.values],
        )

    def transform_attractors(
        self,
        attractors: list[Attractor],
    ) -> list[DiscreteAttractor]:
        return [self.transform_attractor(attractor) for attractor in attractors]


def transform_terms(
    terms: list[SequenceTerm],
    transformer: DiscreteTransformer,
) -> list[DiscreteTerm]:
    return transformer.transform_terms(terms)


def transform_attractors(
    attractors: list[Attractor],
    transformer: DiscreteTransformer,
) -> list[DiscreteAttractor]:
    return transformer.transform_attractors(attractors)

File: src/transformations.rs
use std::sync::Arc;

use anyhow::{Result, bail};
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};

use crate::coordinates::CoordinateSystem;
use crate::discretization::DiscretizationGrid;
use crate::models::{Attractor, Number, SequenceAnalysisResult, SequenceTerm};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DiscretePoint {
    pub value: Number,
    pub coordinates: IndexMap<String, f64>,
    pub indices: Option<IndexMap<String, usize>>,
    pub discrete_coordinates: Option<IndexMap<String, f64>>,
    pub cell_id: Option<usize>,
    pub inside_domain: bool,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DiscreteTerm {
    pub term: SequenceTerm,
    pub point: DiscretePoint,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct DiscreteAttractor {
    pub attractor: Attractor,
    pub points: Vec<DiscretePoint>,
}

#[derive(Clone)]
pub struct DiscreteTransformer {
    pub coordinate_system: Arc<dyn CoordinateSystem>,
    pub grid: DiscretizationGrid,
}

impl DiscreteTransformer {
    pub fn new(coordinate_system: Arc<dyn CoordinateSystem>, grid: DiscretizationGrid) -> Result<Self> {
        let missing: Vec<String> = grid
            .axes()
            .into_iter()
            .filter(|axis| !coordinate_system.axes().contains(&axis.as_str()))
            .collect();
        if !missing.is_empty() {
            bail!(
                "coordinate system does not provide axes: {}",
                missing.join(", ")
            );
        }
        Ok(Self {
            coordinate_system,
            grid,
        })
    }

    pub fn transform_value(&self, value: Number) -> Result<DiscretePoint> {
        let coordinates = self.coordinate_system.coordinates_from(value)?;
        let indices = self.grid.indices_of(&coordinates)?;
        if let Some(indices) = indices {
            let discrete_coordinates = self.grid.centers_of(&indices)?;
            let cell_id = self.grid.cell_id(&indices)?;
            Ok(DiscretePoint {
                value,
                coordinates,
                indices: Some(indices),
                discrete_coordinates: Some(discrete_coordinates),
                cell_id: Some(cell_id),
                inside_domain: true,
            })
        } else {
            Ok(DiscretePoint {
                value,
                coordinates,
                indices: None,
                discrete_coordinates: None,
                cell_id: None,
                inside_domain: false,
            })
        }
    }

    pub fn transform_terms(&self, terms: &[SequenceTerm]) -> Result<Vec<DiscreteTerm>> {
        terms
            .iter()
            .map(|term| {
                Ok(DiscreteTerm {
                    term: term.clone(),
                    point: self.transform_value(term.value)?,
                })
            })
            .collect()
    }

    pub fn transform_attractor(&self, attractor: &Attractor) -> Result<DiscreteAttractor> {
        let points = attractor
            .values
            .iter()
            .map(|value| self.transform_value(*value))
            .collect::<Result<Vec<_>>>()?;
        Ok(DiscreteAttractor {
            attractor: attractor.clone(),
            points,
        })
    }

    pub fn transform_attractors(&self, attractors: &[Attractor]) -> Result<Vec<DiscreteAttractor>> {
        attractors
            .iter()
            .map(|attractor| self.transform_attractor(attractor))
            .collect()
    }
}

#[derive(Clone)]
pub struct DiscreteAnalysisResult {
    pub sequence_result: SequenceAnalysisResult,
    pub transformer: DiscreteTransformer,
    pub terms: Vec<DiscreteTerm>,
    pub attractors: Vec<DiscreteAttractor>,
    pub config: Option<crate::config::ExplorerConfig>,
}

impl DiscreteAnalysisResult {
    pub fn from_analysis(
        sequence_result: SequenceAnalysisResult,
        transformer: DiscreteTransformer,
        config: Option<crate::config::ExplorerConfig>,
    ) -> Result<Self> {
        let terms = transformer.transform_terms(&sequence_result.terms)?;
        let attractors = transformer.transform_attractors(&sequence_result.attractors)?;
        Ok(Self {
            sequence_result,
            transformer,
            terms,
            attractors,
            config,
        })
    }
}

pub fn transform_terms(terms: &[SequenceTerm], transformer: &DiscreteTransformer) -> Result<Vec<DiscreteTerm>> {
    transformer.transform_terms(terms)
}

pub fn transform_attractors(
    attractors: &[Attractor],
    transformer: &DiscreteTransformer,
) -> Result<Vec<DiscreteAttractor>> {
    transformer.transform_attractors(attractors)
}
